File: content/tracks/dsa-python/concepts/19-dynamic-programming-2d/dynamic-programming-2d_test.py
import sys
from stub import unique_paths, longest_common_subsequence, edit_distance

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# unique_paths tests
result = unique_paths(3, 7)
test("paths: 3x7", result == 28, f"got {result}, want 28")

result = unique_paths(3, 2)
test("paths: 3x2", result == 3, f"got {result}, want 3")

result = unique_paths(1, 1)
test("paths: 1x1", result == 1, f"got {result}, want 1")

result = unique_paths(1, 5)
test("paths: 1xN (single row)", result == 1, f"got {result}, want 1")

result = unique_paths(5, 1)
test("paths: Mx1 (single col)", result == 1, f"got {result}, want 1")

result = unique_paths(2, 2)
test("paths: 2x2", result == 2, f"got {result}, want 2")

result = unique_paths(4, 4)
test("paths: 4x4", result == 20, f"got {result}, want 20")

# longest_common_subsequence tests
result = longest_common_subsequence("abcde", "ace")
test("lcs: abcde/ace", result == 3, f"got {result}, want 3")

result = longest_common_subsequence("abc", "abc")
test("lcs: identical strings", result == 3, f"got {result}, want 3")

result = longest_common_subsequence("abc", "def")
test("lcs: no common", result == 0, f"got {result}, want 0")

result = longest_common_subsequence("", "abc")
test("lcs: empty first string", result == 0, f"got {result}, want 0")

result = longest_common_subsequence("abc", "")
test("lcs: empty second string", result == 0, f"got {result}, want 0")

result = longest_common_subsequence("oxcpqrsvwf", "shmtulqrypy")
test("lcs: longer strings", result == 2, f"got {result}, want 2")

# edit_distance tests
result = edit_distance("horse", "ros")
test("edit: horse->ros", result == 3, f"got {result}, want 3")

result = edit_distance("intention", "execution")
test("edit: intention->execution", result == 5, f"got {result}, want 5")

result = edit_distance("", "abc")
test("edit: empty->abc", result == 3, f"got {result}, want 3")

result = edit_distance("abc", "")
test("edit: abc->empty", result == 3, f"got {result}, want 3")

result = edit_distance("abc", "abc")
test("edit: identical", result == 0, f"got {result}, want 0")

result = edit_distance("a", "b")
test("edit: single char replace", result == 1, f"got {result}, want 1")

result = edit_distance("kitten", "sitting")
test("edit: kitten->sitting", result == 3, f"got {result}, want 3")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)

File: content/tracks/dsa-javascript/concepts/19-dynamic-programming-2d/dynamic-programming-2d_test.js
const assert = require("assert");
const { uniquePaths, longestCommonSubsequence, editDistance } = require("./stub");

let passed = 0, failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e) {
    console.log(`  FAIL: ${name} — ${e.message}`);
    failed++;
  }
}

// uniquePaths
test("uniquePaths: 3x7", () => {
  assert.strictEqual(uniquePaths(3, 7), 28);
});
test("uniquePaths: 3x2", () => {
  assert.strictEqual(uniquePaths(3, 2), 3);
});
test("uniquePaths: 1x1", () => {
  assert.strictEqual(uniquePaths(1, 1), 1);
});
test("uniquePaths: 1xN — only one path (go right)", () => {
  assert.strictEqual(uniquePaths(1, 5), 1);
});
test("uniquePaths: Mx1 — only one path (go down)", () => {
  assert.strictEqual(uniquePaths(5, 1), 1);
});
test("uniquePaths: 2x2", () => {
  assert.strictEqual(uniquePaths(2, 2), 2);
});
test("uniquePaths: 3x3", () => {
  assert.strictEqual(uniquePaths(3, 3), 6);
});

// longestCommonSubsequence
test("LCS: 'abcde' vs 'ace' = 3", () => {
  assert.strictEqual(longestCommonSubsequence("abcde", "ace"), 3);
});
test("LCS: identical strings", () => {
  assert.strictEqual(longestCommonSubsequence("abc", "abc"), 3);
});
test("LCS: no common characters", () => {
  assert.strictEqual(longestCommonSubsequence("abc", "def"), 0);
});
test("LCS: one empty string", () => {
  assert.strictEqual(longestCommonSubsequence("abc", ""), 0);
  assert.strictEqual(longestCommonSubsequence("", "abc"), 0);
});
test("LCS: both empty", () => {
  assert.strictEqual(longestCommonSubsequence("", ""), 0);
});
test("LCS: 'bl' vs 'yby' = 1", () => {
  assert.strictEqual(longestCommonSubsequence("bl", "yby"), 1);
});
test("LCS: 'bsbininm' vs 'jmjkbkjkv' = 2", () => {
  assert.strictEqual(longestCommonSubsequence("bsbininm", "jmjkbkjkv"), 2);
});

// editDistance
test("editDistance: 'horse' to 'ros' = 3", () => {
  assert.strictEqual(editDistance("horse", "ros"), 3);
});
test("editDistance: 'intention' to 'execution' = 5", () => {
  assert.strictEqual(editDistance("intention", "execution"), 5);
});
test("editDistance: identical strings = 0", () => {
  assert.strictEqual(editDistance("abc", "abc"), 0);
});
test("editDistance: empty to non-empty = length", () => {
  assert.strictEqual(editDistance("", "abc"), 3);
});
test("editDistance: non-empty to empty = length", () => {
  assert.strictEqual(editDistance("abc", ""), 3);
});
test("editDistance: both empty = 0", () => {
  assert.strictEqual(editDistance("", ""), 0);
});
test("editDistance: single char insert", () => {
  assert.strictEqual(editDistance("a", "ab"), 1);
});
test("editDistance: single char replace", () => {
  assert.strictEqual(editDistance("a", "b"), 1);
});
test("editDistance: 'kitten' to 'sitting' = 3", () => {
  assert.strictEqual(editDistance("kitten", "sitting"), 3);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
